x-pack/plugins/actions/server/routes/get_oauth_access_token.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema, TypeOf } from '@kbn/config-schema';
+import { IRouter, Logger } from '@kbn/core/server';
+import { ILicenseState } from '../lib';
+import { INTERNAL_BASE_ACTION_API_PATH } from '../../common';
+import { ActionsRequestHandlerContext } from '../types';
+import { verifyAccessAndContext } from './verify_access_and_context';
+import { ActionsConfigurationUtilities } from '../actions_config';
+import {
+  getOAuthJwtAccessToken,
+  GetOAuthJwtConfig,
+  GetOAuthJwtSecrets,
+} from '../builtin_action_types/lib/get_oauth_jwt_access_token';
+import {
+  getOAuthClientCredentialsAccessToken,
+  GetOAuthClientCredentialsConfig,
+  GetOAuthClientCredentialsSecrets,
+} from '../builtin_action_types/lib/get_oauth_client_credentials_access_token';
+
+const oauthJwtBodySchema = schema.object({
+  tokenUrl: schema.string(),
+  config: schema.object({
+    clientId: schema.string(),
+    jwtKeyId: schema.string(),
+    userIdentifierValue: schema.string(),
+  }),
+  secrets: schema.object({
+    clientSecret: schema.string(),
+    privateKey: schema.string(),
+    privateKeyPassword: schema.maybe(schema.string()),
+  }),
+});
+
+type OAuthJwtParams = TypeOf<typeof oauthJwtBodySchema>;
+
+const oauthClientCredentialsBodySchema = schema.object({
+  tokenUrl: schema.string(),
+  scope: schema.string(),
+  config: schema.object({
+    clientId: schema.string(),
+    tenantId: schema.string(),
+  }),
+  secrets: schema.object({
+    clientSecret: schema.string(),
+  }),
+});
+
+type OAuthClientCredentialsParams = TypeOf<typeof oauthClientCredentialsBodySchema>;
+
+const bodySchema = schema.object({
+  type: schema.oneOf([schema.literal('jwt'), schema.literal('client')]),
+  options: schema.conditional(
+    schema.siblingRef('type'),
+    schema.literal('jwt'),
+    oauthJwtBodySchema,
+    oauthClientCredentialsBodySchema
+  ),
+});
+
+export const getOAuthAccessToken = (
+  router: IRouter<ActionsRequestHandlerContext>,
+  licenseState: ILicenseState,
+  logger: Logger,
+  configurationUtilities: ActionsConfigurationUtilities
+) => {
+  router.post(
+    {
+      path: `${INTERNAL_BASE_ACTION_API_PATH}/connector/_oauth_access_token`,
+      validate: {
+        body: bodySchema,
+      },
+    },
+    router.handleLegacyErrors(
+      verifyAccessAndContext(licenseState, async function (context, req, res) {
+        const { type, options } = req.body;
+
+        try {
+          configurationUtilities.ensureUriAllowed(options.tokenUrl);
+        } catch (err) {
+          return res.badRequest({ body: err.message });
+        }
```

Comment:
Above and beyond this hostname validation, can we check to ensure that the URL contains _a hostname_ (it should never be a local path), and can we validate the URL scheme to the minimum that's needed (we should only need `https` right?)

See [this awesome blog post](https://www.acunetix.com/blog/articles/server-side-request-forgery-vulnerability/) for more information about mitigating SSRF.

Reply:
Added the recommended additional validation in this commit: https://github.com/elastic/kibana/pull/131171/commits/52b94de3fa3474a5a4c21b604335195007c66677